Add OANDA RestV20 Historical Instrument Reader

davidandreoletti opened the pull request.

davidandreoletti commented:
@femtotrader I create a new Pull Request for several reasons:
- The OANDA API used to fetch the data has changed from OANDA REST API: (`oandapy` dependency) to OANDA REST V20 API (`oandapyV20` dependency) because there is no intention from OANDA to release `oandapy` on PyPy :( See https://github.com/oanda/oandapy/issues/39#issuecomment-258894179

I will have a look at implementing some of your suggestions below. 

```
Thanks @davidandreoletti

a) But you could simply give this PR a title starting with WIP (work in progress) and still push there.
b) In order to provide a better user API experience granularity could be a Pandas offset shortcut string (like '5Min', '5H' ... in resample function ).
c) And it could be convert to a string compatible with Oanda API.
d) We could have a parameter named freq for DataReader function (it's role is quite similar to pandas.date_range parameter)

See #199

I'd convert freq (string) to Pandas DateOffset
http://stackoverflow.com/questions/27368265/convert-freq-string-to-dateoffset-in-pandas

and have a dict to convert Pandas DateOffset to Oanda granularity string.

Ask me if you don't understand what I mean (because I'm not english native speaker) so I may miss something.
```

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -25,24 +25,78 @@ class OANDARestHistoricalInstrumentReader(_BaseReader):
+            "1T":"M1",
+            "1min":"M1",
+            "2T":"M2",
+            "2min":"M2",
```
When you define `SUPPORTED_OFFSET_ALIASES`

You should have Pandas Offset as key not string shortcut as key.
Because it urge you to repeat several times some values.

```python
In [1]: from pandas.tseries.frequencies import to_offset

In [2]: to_offset("1Min")
Out[2]: <Minute>

In [3]: offset = to_offset("1Min")

In [4]: offset
Out[4]: <Minute>

In [5]: type(offset)
Out[5]: pandas.tseries.offsets.Minute
```

davidandreoletti replied:
@femtotrader OANDA REST v20 only accept predefined freqencies. Eg: 1T is supported since M1 is an accepted value for OANDA REST v20's endpoint however 7T -> M7 is not supported since M7 is not a `granularity` value accepted by the endpoint.

Does it still make sense to have `pandas.tseries.offsets.XXXX object` instead of hardcoded Panda Offset aliases ?

femtotrader replied:
On the other side I don't think it's a problem if user pass "freq=7T" which will be convert to `<7 * Minutes>` but it won't be in `SUPPORTED_OFFSET_ALIASES` dict keys so we can decide here to raise an Exception to inform user that it's no an accepted timeframe.

femtotrader replied:
Should we define keys for "1min", "1MIN", "1Min"...

```python
In [9]: to_offset("1min")
Out[9]: <Minute>

In [10]: to_offset("1Min")
Out[10]: <Minute>

In [11]: to_offset("1MIN")
Out[11]: <Minute>
```

femtotrader replied:
I don't think so

davidandreoletti replied:
@femtotrader Ok. So i am thinking of accepting a string or DateOffset for `freq` param

How can a DateOffset object be converted back to string ?  Eg: 
`
In[1]d = to_offset("1min") 
In[2] ds = to_offset_string(d)
Out[2] print(ds) # Displays "1min"
`

femtotrader replied:
I'm not sure a DateOffset can be convert back to a "normalized" offset string

```
In [20]: offset = to_offset("7T")

In [21]: offset.
                 offset.apply       offset.delta       offset.kwds        offset.nanos       offset.rollback
                 offset.apply_index offset.freqstr     offset.n           offset.normalize   offset.rollforward
                 offset.copy        offset.isAnchored  offset.name        offset.onOffset    offset.rule_code
```

but it could be an interesting feature to have in Pandas

```python
In [18]: offset.rule_code
Out[18]: 'T'

In [19]: offset.rule_code
Out[19]: 'T'

In [20]: offset = to_offset("7T")

In [21]: offset.n
Out[21]: 7

In [22]: offset.rule_code
Out[22]: 'T'

In [23]: offset = to_offset("7Min")

In [24]: offset.n
Out[24]: 7

In [25]: offset.rule_code
Out[25]: 'T'
```

a quick and dirty approach could be

```
In [47]: def normalized_offset_string(offset):
    ...:      if offset.n == 1:
    ...:          return offset.rule_code
    ...:      else:
    ...:          return str(offset.n) + offset.rule_code
```

femtotrader commented:
There is in fact an attribute to get normalized string from DateOffset it's named: `freqstr`

```python
In [64]: to_offset("7Min").freqstr
Out[64]: '7T'

In [65]: to_offset("7MIN").freqstr
Out[65]: '7T'

In [66]: to_offset("7min").freqstr
Out[66]: '7T'
```

femtotrader commented:
I'm sorry I don't know "all" Pandas ;-)

davidandreoletti commented:
@femtotrader That's fine :) I only expect you to know the answers to the questions I ask :P

davidandreoletti commented:
@femtotrader My PR is done.

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,308 @@
+            "5T":"M5",
+            "5min":"M5",
+            "15T":"M15",
+            "15min":"M15",
```
`freqstr` DateOffset attribute should avoid to repeat these keys

```python
In [67]: to_offset("15min").freqstr
Out[67]: '15T'

In [68]: to_offset("15T").freqstr
Out[68]: '15T'

In [69]: to_offset("15Min").freqstr
Out[69]: '15T'
```

davidandreoletti replied:
@femtotrader Fixed.

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,308 @@
+            current_end_rfc3339 = current_end.strftime(rfc3339)
+
+            if isinstance(freq, DateOffset):
+                offsetString = self._normalized_offset_string(freq)
```
```python
offsetString = freq.freqstr
```

Sorry;-)

davidandreoletti replied:
@femtotrader Fixed.

PS: No worries at all :)

femtotrader commented on pandas_datareader/data.py:
```diff
@@ -56,7 +57,8 @@ def get_quote_google(*args, **kwargs):
 def DataReader(name, data_source=None, start=None, end=None,
-               retry_count=3, pause=0.001, session=None, access_key=None):
+               retry_count=3, pause=0.001, session=None, 
+               custom=None):
```
`access_key` is required for Enigma datareader.

davidandreoletti replied:
@femtotrader Fixed.

femtotrader commented on pandas_datareader/data.py:
```diff
@@ -107,6 +111,13 @@ def DataReader(name, data_source=None, start=None, end=None,
+    oa = DataReader("EUR_USD", data_source="oanda_rest_historical_currency", 
+                    custom={    "accountType"="practice",
+                                    "accountVersion="0"
+                                    "apiToken":"Your private API token" })
```
custom = {
                        "accountType": "practice",
                        "accountVersion": "0",
                        "apiToken": "Your private API token"
                    })

davidandreoletti replied:
@femtotrader Fixed

femtotrader commented:
You need to run locally

```
flake8 --ignore E501 pandas_datareader
```

to fix some Flake8 errors

davidandreoletti commented:
@femtotrader PR done :)

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,303 @@
+                      "bid.c": "bid.close",
+                  })
+
+        df = df.set_index('date')
```
A DataFrame with a MultiIndex as columns will be a great feature to have
http://pandas.pydata.org/pandas-docs/stable/advanced.html

Most other DataReader with candlestick data are using; `Open`, `High`, `Low`, `Close` column name (not `open`, `high`, `low`, `close`)

Here is an example to show how a DataFrame with MultiIndexed columns works
```python
In: periods = 10
In: idx = pd.date_range("2010-01-01", freq="D", periods=10)
In: columns = [np.array(['Mid', 'Mid', 'Mid', 'Mid', 'Ask', 'Ask', 'Ask', 'Ask', 'Bid', 'Bid', 'Bid', 'Bid']),
              np.array(['Open', 'High', 'Low', 'Close', 'Open', 'High', 'Low', 'Close', 'Open', 'High', 'Low', 'Close'])]

In: df = pd.DataFrame(np.random.randn(periods, 12), columns=columns, index=idx)

In: df
Out[59]: 
                 Mid                                     Ask            \
                Open      High       Low     Close      Open      High   
2010-01-01 -0.357908  0.638564 -0.808755  0.029079  0.150620  0.282146   
2010-01-02  1.015372 -0.624254  0.353841  1.491935 -0.322171  0.353117   
2010-01-03  0.051334  0.516878 -1.769581  0.812352 -0.528826  0.677380   
2010-01-04 -1.681174  0.411583 -1.032651  1.924580  0.162397 -0.422782   
2010-01-05  0.512275 -0.112695 -0.705641  0.715283  0.446572 -1.428909   
2010-01-06 -1.197475 -0.025064 -0.997523  1.252269  0.217094 -1.086155   
2010-01-07 -0.935145 -0.446475 -1.220699 -1.135570  0.369933  1.695358   
2010-01-08 -0.259295  0.164924  0.371293 -0.136309  2.280143 -1.256588   
2010-01-09 -0.392548  1.762788 -1.656464  1.548067 -0.333061  0.226948   
2010-01-10 -0.431441  0.674132 -1.381039  0.892813  0.793012 -0.755067   

                                     Bid                                
                 Low     Close      Open      High       Low     Close  
2010-01-01 -1.313252 -1.179413  1.873345 -0.719676 -2.595192  1.182884  
2010-01-02 -0.670126  0.421320  0.318873 -0.756884  1.189023 -0.785939  
2010-01-03 -1.027978  0.609855  0.877085 -0.767293 -1.859567  0.662125  
2010-01-04  1.049496 -1.062435  0.577158 -0.068621 -0.080246 -0.786441  
2010-01-05 -3.193001 -0.540472  2.350964 -1.983940  2.278458 -0.783962  
2010-01-06 -0.101567 -0.145842 -1.600600  0.098544  0.133307  0.262338  
2010-01-07 -0.263570 -1.012718 -0.579255  1.360744  1.661800 -2.185659  
2010-01-08 -2.260959 -1.143676 -0.900669  0.721042 -0.421265 -0.716886  
2010-01-09 -1.229502 -0.442861 -0.767872  0.734038 -0.403405  0.046703  
2010-01-10  1.522349  0.463781  0.522694 -0.700728 -0.726378 -1.144720 

df["Ask"]
Out[60]: 
                Open      High       Low     Close
2010-01-01  0.150620  0.282146 -1.313252 -1.179413
2010-01-02 -0.322171  0.353117 -0.670126  0.421320
2010-01-03 -0.528826  0.677380 -1.027978  0.609855
2010-01-04  0.162397 -0.422782  1.049496 -1.062435
2010-01-05  0.446572 -1.428909 -3.193001 -0.540472
2010-01-06  0.217094 -1.086155 -0.101567 -0.145842
2010-01-07  0.369933  1.695358 -0.263570 -1.012718
2010-01-08  2.280143 -1.256588 -2.260959 -1.143676
2010-01-09 -0.333061  0.226948 -1.229502 -0.442861
2010-01-10  0.793012 -0.755067  1.522349  0.463781

df["Ask"]["Open"]
Out[61]: 
2010-01-01    0.150620
2010-01-02   -0.322171
2010-01-03   -0.528826
2010-01-04    0.162397
2010-01-05    0.446572
2010-01-06    0.217094
2010-01-07    0.369933
2010-01-08    2.280143
2010-01-09   -0.333061
2010-01-10    0.793012
Freq: D, Name: Open, dtype: float64
```

davidandreoletti replied:
Fixed: Use `Open` instead `open`, etc ...

I will look at A DataFrame with a MultiIndex later this week.

davidandreoletti replied:
Fixed: Added MultiIndex support. Thanks for the suggestion @femtotrader !

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,303 @@
+            self.access_credential['apiToken'] = ""
+
+    def read(self):
+        dfs = {}
```
using an OrderedDict will keep order of symbols in this Panel

    from pandas.compat import OrderedDict

davidandreoletti replied:
Fixed

femtotrader commented on pandas_datareader/tests/test_oandarest.py:
```diff
@@ -0,0 +1,74 @@
+    def test_oanda_historical_currencypair(self):
+        start = "2014-03-19T09:00:00Z"
+        end = "2014-03-21T9:00:00Z"
+        symbols = ["EUR_USD"]
```
Does it works with simply

    symbols = "EUR_USD"

?

femtotrader replied:
Nice.
Great job on this PR!

femtotrader commented on pandas_datareader/tests/test_oandarest.py:
```diff
@@ -0,0 +1,74 @@
+    def test_oanda_historical_currencypair2(self):
+        start = "2014-03-19T09:00:00Z"
+        end = "2014-03-21T09:00:00Z"
+        symbols = ["EUR_USD"]
```
A test with several symbols to ensure that we get a Panel will be a good idea.

davidandreoletti replied:
Fixed

davidandreoletti commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,303 @@
+        "6H": "H6",
+        "8H": "H8",
+        "12H": "H12",
+        "1D": "D",
```
Fixed

davidandreoletti commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,303 @@
+        "8H": "H8",
+        "12H": "H12",
+        "1D": "D",
+        "1W": "W",
```
Fixed

davidandreoletti commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,303 @@
+        "12H": "H12",
+        "1D": "D",
+        "1W": "W",
+        "1M": "M"
```
Fixed

davidandreoletti commented on pandas_datareader/tests/test_oandarest.py:
```diff
@@ -0,0 +1,74 @@
+            pn = OANDARestHistoricalInstrumentReader(
+                symbols=symbols,
+                start=start, end=end,
+                freq="S5",
```
Fixed

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,313 @@
+                    8H  ->  8 hour candlesticks, day alignment
+                    12H -> 12 hour candlesticks, day alignment
+                    1D  ->  1 day candlesticks, day alignment
+                    1W  ->  1 week candlesticks, aligned to start of week
```
1W -> W

davidandreoletti replied:
Fixed

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,313 @@
+                    6H  ->  6 hour candlesticks, day alignment
+                    8H  ->  8 hour candlesticks, day alignment
+                    12H -> 12 hour candlesticks, day alignment
+                    1D  ->  1 day candlesticks, day alignment
```
1D -> D

davidandreoletti replied:
Fixed

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,313 @@
+                    12H -> 12 hour candlesticks, day alignment
+                    1D  ->  1 day candlesticks, day alignment
+                    1W  ->  1 week candlesticks, aligned to start of week
+                    1M  ->  1 month candlesticks, aligned to first day of the month
```
1M -> M

davidandreoletti replied:
Fixed

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,313 @@
+import pandas as pd
```
import os

is required

davidandreoletti commented:
```
OANDA_API_TOKEN is now set on Travis.

I'm not sure but we probably need to export this environment variable in .travis.yml

export OANDA_API_TOKEN=$OANDA_API_TOKEN
```

The initial API TOKEN I committed was a mistake and it has now been revoked. So I will not provide an API TOKEN. Alternatively, someone (else) can provide an OANDA API TOKEN, encrypt it and then store it as encrypted env variable (supported by Travis).

davidandreoletti commented:
@femtotrader PR done.

femtotrader commented:
I add my API key in Travis. But I think you need to add

    export OANDA_API_TOKEN=$OANDA_API_TOKEN

in `.travis.yml`

davidandreoletti commented:
@femtotrader OANDA_API_TOKEN env export from travis :)

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,376 @@
+        if 'accountType' not in access_credential:
+            self.access_credential['accountType'] = "practice"
+
+        if 'apiToken' not in access_credential:
```
Same here

    if 'apiToken' not in access_credential:

should be

    if 'apiToken' not in self.access_credential:

davidandreoletti replied:
@femtotrader fixed

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,376 @@
+        if access_credential is None:
+            self.access_credential = {}
+
+        if 'accountType' not in access_credential:
```
Here it's problematic

So

    if 'accountType' not in access_credential:

should be

    if 'accountType' not in self.access_credential:

because

```python
pdr.DataReader("EUR_USD", "oanda_historical_currency", "2014-03-19T09:00:00Z", "2014-03-20T09:00:00Z")
```

raises

    TypeError: argument of type 'NoneType' is not iterable

davidandreoletti replied:
@femtotrader fixed

femtotrader commented on pandas_datareader/oandarest.py:
```diff
@@ -0,0 +1,376 @@
+            self.candleFormat = OANDARestHistoricalInstrumentReader.DEFAULT_CANDLE_FORMAT
+
+        self.access_credential = access_credential
+        if access_credential is None:
```
if access_credential is None:

should be

    if self.access_credential is None:

(but that's not problematic here)

davidandreoletti replied:
@femtotrader fixed

femtotrader commented:
You might also transpose panel to have same dimension order than other datareader such as yahoo datareader


    Items axis: prices (Open, High, Low, Close), Volume
    Major_axis axis: timestamp
    Minor_axis axis: symbol

Here is an example:

```python
In [4]: pdr.DataReader(["IBM", "GOOG"], "yahoo")
Out[4]:
<class 'pandas.core.panel.Panel'>
Dimensions: 6 (items) x 1729 (major_axis) x 2 (minor_axis)
Items axis: Open to Adj Close
Major_axis axis: 2010-01-04 00:00:00 to 2016-11-11 00:00:00
Minor_axis axis: GOOG to IBM

In [5]: data = pdr.DataReader(["EUR_USD", "USD_JPY"], "oanda_historical_currency", "2014-03-19T09:00:00Z", "2014-03-20T09:00:00Z")
request https://api-fxpractice.oanda.com/v3/instruments/EUR_USD/candles failed [400,{"errorMessage":"The results returned satisfying the query exceeds the maximum size"}]
request https://api-fxpractice.oanda.com/v3/instruments/USD_JPY/candles failed [400,{"errorMessage":"The results returned satisfying the query exceeds the maximum size"}]

In [6]: data
Out[6]:
<class 'pandas.core.panel.Panel'>
Dimensions: 2 (items) x 15438 (major_axis) x 6 (minor_axis)
Items axis: EUR_USD to USD_JPY
Major_axis axis: 2014-03-19 09:00:00 to 2014-03-20 08:59:55
Minor_axis axis: (Mid, Close) to (Mid, Complete)
```

I also noticed that you are also returning a Panel when 1 symbol is given.

```python
In [7]: pdr.DataReader("EUR_USD", "oanda_historical_currency", "2014-03-19T09:00:00Z", "2014-03-20T09:00:00Z")
request https://api-fxpractice.oanda.com/v3/instruments/EUR_USD/candles failed [400,{"errorMessage":"The results returned satisfying the query exceeds the maximum size"}]
Out[7]:
<class 'pandas.core.panel.Panel'>
Dimensions: 1 (items) x 12125 (major_axis) x 6 (minor_axis)
Items axis: EUR_USD to EUR_USD
Major_axis axis: 2014-03-19 09:00:00 to 2014-03-20 08:59:55
Minor_axis axis: (Mid, Close) to (Mid, Complete)
```

I would return either
- a DataFrame with hierarchical columns
- or a Panel but symbols shouldn't be in axis (there's only one symbol)

femtotrader commented:
I also noticed that

    data=pdr.DataReader("EUR_USD", "oanda_historical_currency", "2014-03-19T09:00:00Z", "2014-03-20T09:00:00Z")

    data["EUR_USD"]["Mid"]

displays

```
In [36]: data["EUR_USD"]["Mid"]
Out[36]:
                       Close     High      Low     Open Volume Complete
Date
2014-03-19 09:00:00  1.39142  1.39144  1.39141  1.39141      5     True
2014-03-19 09:00:05  1.39142  1.39142  1.39142  1.39142      4     True
2014-03-19 09:00:10  1.39143  1.39145  1.39143  1.39145      3     True
2014-03-19 09:00:15  1.39148  1.39148  1.39144  1.39146      6     True
2014-03-19 09:00:20  1.39145  1.39149  1.39145  1.39148     10     True
...                      ...      ...      ...      ...    ...      ...
2014-03-20 08:59:25  1.38251  1.38251  1.38247  1.38247      7     True
2014-03-20 08:59:40  1.38249  1.38249  1.38249  1.38249      1     True
2014-03-20 08:59:45  1.38250  1.38250  1.38247  1.38250      6     True
2014-03-20 08:59:50  1.38248  1.38251  1.38248  1.38251      2     True
2014-03-20 08:59:55  1.38226  1.38248  1.38225  1.38248     36     True

[12125 rows x 6 columns]
```

but both `data["EUR_USD"]["Ask"]` and `data["EUR_USD"]["Bid"]` raises a `KeyError`

femtotrader commented:
Good work @davidandreoletti 
but there is still a hint

See

```python
In [45]: panel = pdr.DataReader(["GOOG", "IBM"], "yahoo")

In [46]: panel
Out[46]:
<class 'pandas.core.panel.Panel'>
Dimensions: 6 (items) x 1743 (major_axis) x 2 (minor_axis)
Items axis: Open to Adj Close
Major_axis axis: 2010-01-04 00:00:00 to 2016-12-02 00:00:00
Minor_axis axis: GOOG to IBM

In [47]: panel["Open","2016-11-15","GOOG"]
Out[47]: 746.96997099999999
```

but with `oanda_historical_currency` source

```python
In [48]: data = pdr.DataReader("EUR_USD", "oanda_historical_currency", "2014-03-
    ...: 19T09:00:00Z", "2014-03-20T09:00:00Z")
request https://api-fxpractice.oanda.com/v3/instruments/EUR_USD/candles failed [400,{"errorMessage":"The results returned satisfying the query exceeds the maximum size"}]
In [49]: data
Out[49]:
<class 'pandas.core.panel.Panel'>
Dimensions: 6 (items) x 12125 (major_axis) x 1 (minor_axis)
Items axis: (Mid, Close) to (Mid, Complete)
Major_axis axis: 2014-03-19 09:00:00 to 2014-03-20 08:59:55
Minor_axis axis: EUR_USD to EUR_USD

In [50]: data["Mid"]["Close","2014-03-19 09:00:00","EUR_USD"]
Out[50]:
Date
2014-03-19 09:00:00    1.39142
Name: EUR_USD, dtype: object
```

We are getting a `pandas.core.series.Series`

Same with

```python
In [60]: data.loc[("Mid", "Close"),"2014-03-19 09:00:00","EUR_USD"]
Out[60]:
Date
2014-03-19 09:00:00    1.39142
Name: EUR_USD, dtype: object
```

I don't understand why we are not just getting a value

davidandreoletti commented:
@femtotrader 

A) 
> I also noticed that you are also returning a Panel when 1 symbol is given.
> I would return either
>  - a DataFrame with hierarchical columns
>  - or a Panel but symbols shouldn't be in axis (there's only one symbol)

To be honest, I don't see an issue with having a panel with an axis having a single symbol. I figured returning the same data structure regardless of the number of currency_pair fetched (only the number of elements within a dimension would change) would actually be a nice properties:
- easy to count the number of currencies available
- easy to enumerate them (or lack of it if the panel is empty)

I would not mind an explanation about why either one of your suggested returned data structure would be better.

B) 

> I also noticed that
>    data=pdr.DataReader("EUR_USD", "oanda_historical_currency", "2014-03-19T09:00:00Z", "2014-03-20T09:00:00Z")
> returned
>  Date                  Close     High      Low     Open Volume Complete
>   2014-03-19 09:00:00  1.39142  1.39144  1.39141  1.39141      5     True
> while data["EUR_USD"]["Ask"] and data["EUR_USD"]["Bid"] raises a KeyError

OANDA's historical data (Close, Open, High, Low) are actually segregated into 3 candle prices: 
- Bid prices (B)
- Ask prices (A)
- Midpoint prices (M)

By default, the reader downloads only Midpoint prices. That's why querying Ask or Bid data raises a KeyError. The reader will by default download all 3 candle prices instead of only the midpoint prices.

C) 

> We are getting a pandas.core.series.Series
> I don't understand why we are not just getting a value

Me neither! I am looking into it.

femtotrader commented:
A)
it's just about consistency with others datareaders.

```python
In [74]: pdr.DataReader(["GOOG", "IBM"], "yahoo")
Out[74]:
<class 'pandas.core.panel.Panel'>
Dimensions: 6 (items) x 1743 (major_axis) x 2 (minor_axis)
Items axis: Open to Adj Close
Major_axis axis: 2010-01-04 00:00:00 to 2016-12-02 00:00:00
Minor_axis axis: GOOG to IBM

In [75]: pdr.DataReader("GOOG", "yahoo")
Out[75]:
                  Open        High         Low       Close    Volume  \
Date
2010-01-04  626.951088  629.511067  624.241073  626.751061   3927000
2010-01-05  627.181073  627.841071  621.541045  623.991055   6031900
2010-01-06  625.861078  625.861078  606.361042  608.261023   7987100
2010-01-07  609.401025  610.001045  592.651008  594.101005  12876600
2010-01-08  592.000997  603.251034  589.110988  602.021036   9483900
...                ...         ...         ...         ...       ...
2016-11-28  760.000000  779.530029  759.799988  768.239990   2172200
2016-11-29  771.530029  778.500000  768.239990  770.840027   1604500
2016-11-30  770.070007  772.989990  754.830017  758.039978   2365800
2016-12-01  757.440002  759.849976  737.025024  747.919983   2996900
2016-12-02  744.590027  754.000000  743.099976  750.500000   1448700

             Adj Close
Date
2010-01-04  313.062468
2010-01-05  311.683844
2010-01-06  303.826685
2010-01-07  296.753749
2010-01-08  300.709808
...                ...
2016-11-28  768.239990
2016-11-29  770.840027
2016-11-30  758.039978
2016-12-01  747.919983
2016-12-02  750.500000

[1743 rows x 6 columns]
```

but I understand your POV.

B)
ok

C)

On my side

```python
In [1]:  import pandas_datareader as pdr
In [2]: data = pdr.DataReader("EUR_USD", "oanda_historical_currency", "2014-03-19T09:00:00Z", "2014-03-20T09:00:00Z")
In [3]: type(data["Ask"]["Close"]["EUR_USD"]["2014-03-19 09:05:00"])
Out[15]: pandas.core.series.Series
```

I'd like to understand why I'm getting a Series when you are getting a value!
Maybe you should add a unit test (with `isinstance`)


Some doc is also required.
Maybe showing how to get a DataFrame from Panel will be useful.

After this it should be ready to be merged before a new release of pandas-datareader.

femtotrader commented:
in order to merge you will need to rebase (because of conflicting file `pandas_datareader/data.py`)

davidandreoletti commented:
@femtotrader 

A) I will keep it as is for now and perhaps provide an additional argument for the reader to return one of the 2 suggested data structures when a single symbol is provided.

C)
Still investigating the bug but at least I can reproduce the issue. The issue is related to the frequency used for fetching candles (eg: 5T vs S5)

- test_oanda_historical_currencypair (pandas_datareader.tests.test_oandarest.TestOandaHistoricalInstrumentReader) ... (uses `5T`)
```
<class 'str'>
1.39150
```

- test_oanda_historical_currencypair2 (pandas_datareader.tests.test_oandarest.TestOandaHistoricalInstrumentReader) ... (uses `S5`)
``` 
<class 'pandas.core.series.Series'>
Date
2014-03-19 09:05:00    1.39147
Name: EUR_USD, dtype: object
```

- test_oanda_historical_currencypair3 (pandas_datareader.tests.test_oandarest.TestOandaHistoricalInstrumentReader) ...  (uses `5T`)
```
<class 'str'>
1.39150
```

davidandreoletti commented:
@femtotrader 

> Some doc is also required.
> Maybe showing how to get a DataFrame from Panel will be useful.

Fixed in [005bbbc339975c058f6d2e2e68c910d82f38ca4b](https://github.com/pydata/pandas-datareader/pull/257/commits/005bbbc339975c058f6d2e2e68c910d82f38ca4b)

> in order to merge you will need to rebase (because of conflicting file pandas_datareader/data.py)

Fixed

Only C) need bug fix and the pull request will be good to go!

femtotrader commented:
Doc is here https://github.com/pydata/pandas-datareader/blob/master/docs/source/remote_data.rst

davidandreoletti commented:
@femtotrader Well it turns out the great folks at Pandas got to the bottom of the "Series vs scalar" issue: https://github.com/pandas-dev/pandas/issues/14826

So I will update the doc and considered this PR mergeable.

davidandreoletti commented:
@femtotrader Should pricing values be "string" or float values ? I could not find anything on the doc about it.

femtotrader commented on docs/source/remote_data.rst:
```diff
@@ -542,6 +543,42 @@ Download currency historical rate from `Oanda <https://www.oanda.com/>`__.
+    In [1]: import pandas_datareader.data as web
+    In [2]: start, end = "2016-01-01", "2016-06-01"
+    In [3]: currency = ["EUR_USD"]
+    In [4]: credential["accountType"]="practise"
```
symbols = ["EUR_USD"]
    credential = {}

davidandreoletti replied:
Fixed

femtotrader commented on docs/source/remote_data.rst:
```diff
@@ -542,6 +543,42 @@ Download currency historical rate from `Oanda <https://www.oanda.com/>`__.
+.. code-block:: python
+
+    In [1]: import pandas_datareader.data as web
+    In [2]: start, end = "2016-01-01", "2016-06-01"
```
It may be a good idea to put a shorter duration (to avoid too much request by people who just run doc example)

davidandreoletti replied:
Fixed

femtotrader commented:
prices and volume should be convert to float (or int)

```python
In [35]: pn.dtypes
Out[35]:
Ask  Close       object
     High        object
     Low         object
     Open        object
Bid  Close       object
     High        object
     Low         object
     Open        object
Mid  Close       object
     High        object
     Low         object
     Open        object
     Volume       int64
     Complete      bool
Ask  Volume       int64
     Complete      bool
Bid  Volume       int64
     Complete      bool
dtype: object
```

because it won't be possible with string to calculate sum, mean, it won't be possible to plot prices or volume...

femtotrader commented:
Some people could prefer to have `decimal` instead of `float` because of floating point inaccuracy.
But I think that other DataReader never use `decimal` (so you can do a decimal optional output if you want or just float for prices / int for volume)

davidandreoletti commented on docs/source/remote_data.rst:
```diff
@@ -542,6 +543,42 @@ Download currency historical rate from `Oanda <https://www.oanda.com/>`__.
+                symbols, data_source="oanda_historical_currency",
+                start=start, end=end,
+                access_key=credentials)
+            )
```
Fixed

davidandreoletti commented:
@femtotrader PR updated :)

davidandreoletti commented:
@femtotrader Is OANDA_API_TOKEN now unset in Travis (or revoked) ? CI fails with 401 errors when running the tests for this reader.

ischurov commented:
Hi, could you please comment on the proposed solution to pandas-dev/pandas#14826 ? (See [https://github.com/pandas-dev/pandas/issues/14826#issuecomment-266037484](this comment) and below.)

davidandreoletti commented on docs/source/remote_data.rst:
```diff
@@ -542,6 +543,42 @@ Download currency historical rate from `Oanda <https://www.oanda.com/>`__.
+    In [1]: import pandas_datareader.data as web
+    In [2]: start, end = "2016-01-01", "2016-02-01"
+    In [3]: symbols = ["EUR_USD"]
+    In [4]: credentials["accountType"]="practise"
```
Fixed

davidandreoletti replied:
Fixed

davidandreoletti commented:
@femtotrader PR ready.

femtotrader commented:
I'd be very pleased to merge it... but CI is still failing.

I don't understand why.

Could you try to add simply

        print(self.access_credential)

after

        if 'apiToken' not in self.access_credential:
            self.access_credential['apiToken'] = os.getenv('OANDA_API_TOKEN')
            if self.access_credential['apiToken'] is None:
                raise ValueError(
                    """Please provide an OANDA API token or set the OANDA_API_TOKEN environment variable\n
                    If you do not have an API key, you can get one here: http://developer.oanda.com/rest-live/authentication/""")

You should show my API key (so this approach is not very secure... but enough here because I don't mind if you display it in a Travis build.)

See https://github.com/femtotrader/test

davidandreoletti commented:
@femtotrader The API token is missing. See [output](https://travis-ci.org/pydata/pandas-datareader/jobs/183438980#L883)

femtotrader commented:
That's very odd because I set in Travis

OANDA_API_TOKEN=9b7bafef6696cce74f062e039d1e13f8-cc05f4e830729f4e8beb8ba136903cd6

Put this token in code because I'm fed up with Travis

femtotrader commented:
About request and multithreading you might have a look at
http://stackoverflow.com/questions/38280094/python-requests-with-multithreading

davidandreoletti commented:
@femtotrader 

> Put this token in code because I'm fed up with Travis

I advise you not to. OANDA may revoke this token at best or close your account for some breach (after all those are considered personal info)

> About request and multithreading you might have a look at http://stackoverflow.com/questions/38280094/python-requests-with-multithreading

Python is not well suited for multithreaded code due to [GIL](https://wiki.python.org/moin/GlobalInterpreterLock). I have limited GIL knownlegde and initially thought this was a concern for CPU intensive apps. Turns out it also has implications with IO as per the doc provided. 

This PR uses `oanda-api-v20` to access OANDA's endpoints. So I [lodged a feature request](https://github.com/hootnot/oanda-api-v20/issues/61) to that project.

hootnot commented:
@davidandreoletti 
Use requests_mock to fake the API requests. That way you can bypass the token sharing problem. 
Take a look at the tests in the oandapyV20 wrapper repo: https://github.com/hootnot/oanda-api-v20

davidandreoletti commented:
@hootnot Do you know why the environment variable `OANDA_API_TOKEN`'s value is not visible in unit test run on Travis while in my personal environment it is visible ? If you do, would you mind explaining it to me, please :) 

@femtotrader I don't have access to Travis to investigate the environment variable issue, can you investigate it or grant me temporary access to Travis ?

femtotrader commented:
Pinging @sinhrks @davidastephens @jreback 
Could you help on this ?

hootnot commented:
@davidandreoletti @femtotrader 

regarding a):

This is not how it works. Travis can never know the value of the OANDA_API_TOKEN. If you use it from your shell this works. But the shell Travis uses has the value not set, so it it empty.

If you want to use the/your token in the tests (and I would advise you not to) you need to set it up the way Travis points out in their docs, using encryption. The owner of the repo must do this since it involves the a procedure with keys.
You end up with a section in .travis looking like:
```yaml:
env:
  global:
  - secure: xxxxxxxxxxxxxxxxxxxxxxx
```
But when you do it this way you share a token as a contributor to the repo-owner. You do not want that for both parties:
- as a contributor you can revoke the token and tests would break, so the owner relies on a contributor here
- As owner you have access to a token which was offered on certain conditions by the issueing party

If you want to use the token in the tests this way I suggest the best solution is that the repo-owner obtains a token and does the Travis setup with that token. This way the owner has control and contributors can still run the tests.

If you want to use encrypted environment variables, it is up to to repo-owner to set this up.

I use this encryption setup in https://github.com/hootnot/oanda-api-v20/ to deploy to pypi automatically when tagging a release.

I still think that you should use *requests_mock* to setup your tests

Regarding d) 

I think it is wise to split some things up. The oandarest.py contains a lot of constants: 
- the offset-aliases
- the self.OANDA_* and self.DATAFRAME_*

These could be seperated and that makes them importable in other code also, which is currently not the case. It also reduces the size of oandarest about 10% making it more readable.

gliptak commented:
@davidandreoletti Support for Oanda was removed with #366 Consider closing this issue

femtotrader commented:
@davidandreoletti you did a great work on this PR but
I'm closing because of OANDA restrictions.
